Thoms' changes to sine2020 section (#69)
DanNixon/sine2020_software_guidelines_paper: guidelines_for_collaborative_development_of_sustainable_data_treatment_software.tex

\todo{TODO: mention scientific coordination?}

\section{Software development for data treatment software for the neutron scattering and muon spectroscopy}
\label{SINE2020}

> 

Author: ```suggestion
This section describes guidelines specific to the development of data treatment software for the neutron and muon communities.
```

> 

Author: I think this is important to say

Reviewer: I think it is in the new new version

These specific guidelines were developed as part of the data treatment work package (WP 10) in SINE2020.
SINE2020 was a 4-year project between 18 partner institutions spread across 12 different countries. 

> Seven of the partners contributed to WP10 comprising six European neutron scattering and muon spectroscopy facilities (ESS, FZJ, LLB, ILL, ISIS, PSI) and University of Parma. 

Author: ```suggestion
Seven of the partners contributed to WP10 comprising of the university or Parma and six European neutron scattering and muon spectroscopy facilities (ESS, FZJ, LLB, ILL, ISIS, and PSI). 
```

Reviewer: In the new version all this is omitted. I find it annoying to read. On the other hand, it is also important to show how widely the community has been involved, so if somebody can find an elegant way of showing that would be useful.

Reviewer: committing change to this version

> Technical University of Delft and Technical University of Munich, which both operates a research reactor as well as the HZB research reactor and Technical University of Denmark participated as observers in the work package.   

Author: ```suggestion
Observers of the work package included the Technical University of Delft, the Technical University of Munich and the Technical University of Denmark.   
```

Reviewer: Looks ok to me for this version, but again this has been removed from new version

> The first deliverable (D10.1) was to write the standards and guidelines for software developed in SINE2020 and the community as such with the aim of fostering development of sustainable, collaborative and interoperable software. 

Author: ```suggestion
The first deliverable (D10.1) was to write a report on standards and guidelines for software development for
SINE2020. 
These guidelines were then used by the community to develop sustainable, collaborative and interoperable software. 
```

Reviewer: ok for this version

> To this end, the domain of scientific computing was considered in addition to an analysis of five programs considered in SINE2020. 

Author: this does not add anything
```suggestion
```

Reviewer: agree. I think the point was supposed be to that the choices made reflects what is being used in scientific computing / data science more generally. In fact, some of the guidelines can be found already in the proposal. In the new version I just mention that many other programs (almost) adhere to the guidelines and add a few examples that needs to be checked.

> A brief description of those programs are given below followed by a discussion of specific guidelines for the neutron scattering and muon spectroscopy domain. 

Author: ```suggestion
A brief description of the software programs are given below along with a discussion of specific guidelines for the neutron scattering and muon spectroscopy communities. 
```

Reviewer: ok

> \subsection{Analysed data treatment software for neutron scattering and muon spectroscopy}

Author: this is not a clear title

Reviewer: Perhaps not, in any case it does not fit to the new version. The new one is: "Guidelines for development of data treatment software for neutron scattering and muon spectroscopy" which is awfully long. Better ones are welcomed.

> The guidelines for data treatment software for the neutron scattering and muon spectroscopy communities are based on a questionnaire that five existing software projects have completed. Here we give a brief description of those software projects:

Author: ```suggestion
The original SINE2020 guidelines for data treatment software were based on a questionnaire, that was completed by five existing software projects.
```

Reviewer: ok for this version. Again, I am not sure that we need a description of the software at all.

\begin{description}

> \item[BornAgain, www.bornagainproject.org] BornAgain started out as a reimplementation of the software XXX for simulation of a specific scattering technique grazing incidence Small Angle Scattering.

Author: ```suggestion
\item[BornAgain, www.bornagainproject.org] BornAgain provides a framework for calculating the scattering process from a multilayer sample, which may have a rough or smooth interface and can include the effect of nanoparticles.
It is a more generic version of GISAXS by R. Lazzari [J. Appl. Cryst. 35, 406–421 (2002)].
```

Reviewer: thanks!

In the course of SINE2020 it has been extended to other reflectometry techniques.

> The project started nine years ago and have around three dedicated full time developers. 

Author: ```suggestion
The project started nine years ago and is supported by three full time developers. 
```

Reviewer: ok

The software is primarily developed in C++ with a scripting interface in Python and a GUI based on Qt. The software is hosted on GitHub. 

> \item[Mantid, www.mantidproject.org] Mantid is a project initiated by ISIS eleven years ago and since then joined by SNS, ESS, and ILL, with the purpose of providing data reduction software for all their instruments. Besides these facilities Mantid is also being used for specific instruments at other neutron (and muon) sources such as SINQ, FRM-II, and ANSTO. Mantid is developed in C++ and Python with a GUI based on Qt. 

Author: ```suggestion
\item[Mantid, www.mantidproject.org] Mantid is a project initiated by ISIS eleven years ago and now has contributions from the SNS, ESS, and ILL. 
The purpose of Mantid is to provide a single data reduction software for all of the instruments at each of the facilities.
Besides these facilities Mantid is also being used for specific instruments at other neutron (and muon) sources such as SINQ, FRM-II, and ANSTO. 
Mantid is developed using C++ and Python with a Qt based GUI. 
```

Reviewer: ok

> The software is hosted on GitHub under the GPL3 license and the project has about 20 full time developers across all facilities involved.

Author: ```suggestion
The Mantid software is hosted on GitHub under the GPL3 license.
The project has about 20 full time developers spread across each of the partner facilities.
```

Reviewer: ok

> It is by far the largest software project specific for data treatment at neutron and muon sources and has annual developer and user meetings, a XXX. 

Author: ```suggestion
As one of the largest software projects for data treatment at neutron and muon sources, regular feedback is important.
Therefore, Mantid has annual user and developer meetings.
```

Reviewer: ok

> \item[KipTool \& MuhRec, www.neutronimaging.ch] KipTool and MuhRec are analysis software for the neutron imaging technique, where KipTool is used for quantitative analysis and MuhRec for computed tomography and has started as a spare time project for an instrument scientist who spent his time coding when commuting to and from work on the train. 

Author: ```suggestion
\item[KipTool \& MuhRec, www.neutronimaging.ch] KipTool and MuhRec are analysis software for the neutron imaging technique.
KipTool is used for quantitative analysis of imaging data and MuhRec is used for computed tomography.
```

Reviewer: ok

> Since then at most two full-time developers have been dedicated to the project at any given time.

Author: are MuhRec and KipTool one or two projects?

Reviewer: Well, to my understanding they are two different software programs but hosted in the same repository.  However, I don't see any specific kiptool releases (https://github.com/neutronimaging/imagingsuite). We can leave this for Anders K to sort out - I have a meeting with him on Friday.

Author: did we get an answer to this?

Both software programs are originally implemented in C++ with the only user interface being in the form of a GUI based on Qt.
In the course of the SINE2020 project, the software has been migrated to GitHub under the GPL3 license. 

> \item[McStas, www.mcstas.org] McStas is per se not a data treatment software but used for performing instrument simulations.

Author: ```suggestion
\item[McStas, www.mcstas.org] McStas is not a data treatment software, but it is used for performing instrument simulations.
```

Reviewer: ok

> In was included in the analysis due to its widespread usage in the neutron scattering community. 

Author: ```suggestion
These simulations can provide crucial insight into experiments when coupled with data treatment software. 
Hence, it has been included in this analysis as an important tool to the neutron scattering community. 
```

Reviewer: good

> It is also the longest living among those analyzed with more than 20 years of effort behind it but only has one dedicated full-time developer associated to it. 

Author: ```suggestion
McStats started more than 20 years ago and has continued to see new functionality being added. 
However, it only has one dedicated full-time developer. 
```

Reviewer: ok

McStas is developed primarily in C and has recently been migrated to GitHub.
It is licensed under the GPL2 licence. 

> \item[SasView, www.sasview.org] SasView has only had dedicated full time developers when externally funded and has otherwise depended on the good will of scientists putting time into the project.

Author: ```suggestion
\item[SasView, www.sasview.org] SasView has been developed to be a simple and effective set of tools for the analysis of small angle neutron scattering data. 
```

Reviewer: ok

> SasView has had at most two full time developers dedicated to it but has contributions from around fifteen developers from both photon and neutron sources, universities and other research institutions world wide. 

Author: ```suggestion
SasView has had developers from both photon and neutron sources, with around fifteen contributors. 
Typically SasView has two full time developers, which is supplemented by additional developers at code camps (held once or twice a year).

had at most two full time developers dedicated to it but has contributions from around fifteen , universities and other research institutions world wide. 
```

Reviewer: ok

> These developers often contribute during code camps held once or twice per year. 

Author: ```suggestion
```

Reviewer: This is perhaps the most important thing about SasView. I.e. that it for years has been kept alive by having these code camps and with almost nothing happening between. Anyway, removed in new version

> To govern the project, SasView has a management team composed of an instrument scientist from each of the four neutron (and muon) sources: ESS, ISIS, NIST, and SNS. 

Author: ```suggestion
To govern the project, SasView has a management team composed of an instrument scientist from each of the four neutron (and muon) sources (ESS, ISIS, NIST and SNS). 
```

Reviewer: ok

The software is developed in Python with some C and hosted on GitHub under the BSD license. 

> Prior to SINE2020, SasView had a GUI based on WX and no Python interface.

Author: ```suggestion
Prior to SINE2020, SasView had a GUI was based on WX and did not have a Python API.
```

Author: @thomasholmrod Im assuming you mean API when you say interface, so not to confuse it with GUI

Reviewer: In essence it is both a user interface and an API.  Perhaps "Python scripting interface" would be better?

> In the course of SINE2020, it has been equipped with a Python interface as well as a GUI based on Qt (from version 5). 

Author: ```suggestion
In the course of SINE2020, it has been equipped with a Python API as well as a GUI based on Qt (from version 5). 
```

Reviewer: see previous comment

      \item Use a tool for continuous integration and automated testing; such as Travis-CI and Jenkins
      \item Follow a coding standard and use tools that enforces this; such as Pylint (Python) and ClangFormat (C++).
\end{itemize}

The statistics of how guidelines could be applied to software projects are presented in table \ref{summaryTable}.
Some of the projects have a large numbers of developers and/or based at multiple locations.

> As a result the management overheads (e.g., project management boards) are greater to ensure a uniform standard across the project.

Author: ```suggestion
As a result the management overheads (e.g. project management boards) are greater to ensure a uniform standard across the project.
```

Reviewer: oh well, I guess there are different opinions about a ','  

_The consensus seems to be in favor of the comma in American usage; against it in British usage._ (https://www.dailywritingtips.com/comma-after-i-e-and-e-g/)
